### PR TITLE
ISSUE:569: Support Kerberos credentials based Basic Authentication

### LINKED ISSUE #569: Support Kerberos Credentials based Basic Authentication
Currently, we have to authentication handlers
1. Kerberos SPNEGO Authentication Handler
2. Knox based JWT handler.

We are introducing a new authentication handler(KerberosLoginAuthenticationHandler) which allows to authenticate a user based on userId and password he/she provides in https POST request. In the absence of credentials or https or login failure, SPNEGO is attempted just as it's done now. 
The handler expects the credentials in the Authorization header of the request in Base64 encoded bytes of "userId:password" string.

### PR DESCRIPTION
**Pull request by guruchai**
Fixes #569

### CODE REVIEW

**Comment by raju-saravanan** [comment]
@guruchai: Can you create an ISSUE for this and explain what is being done in this PR?

**Comment by guruchai** [comment]
> @guruchai: Can you create an ISSUE for this and explain what is being done in this PR?

Done. Thanks!

File: common-auth/src/test/java/com/hortonworks/registries/auth/server/TestKerberosBasicAuthenticationHandler.java [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+
+        Mockito.when(request.getHeader(KerberosAuthenticator.AUTHORIZATION))
+                .thenReturn(KerberosAuthenticator.NEGOTIATE + " " + token);
+        Mockito.when(request.getServerName()).thenReturn("localhost");
+        Mockito.when(request.getMethod()).thenReturn("GET");
+
+        AuthenticationToken authToken = handler.authenticate(request, response);
+
+        if (authToken != null) {
+            Assert.fail("AuthToken should have been null if the SPNEGO is disabled in KerberosAuthenticationLoginHandler.");
```
Comment (kamalcph) [inline common-auth/src/test/java/com/hortonworks/registries/auth/server/TestKerberosBasicAuthenticationHandler.java]: `KerberosAuthenticationLoginHandler` -> `KerberosBasicAuthenticationHandler`

File: common-auth/src/test/java/com/hortonworks/registries/auth/server/TestKerberosBasicAuthenticationHandler.java [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+        };
+        HttpServletResponse response = Mockito.mock(HttpServletResponse.class);
+
+        Assert.assertNull(handler.authenticate(request, response));
+        Mockito.verify(response).setHeader(KerberosAuthenticator.WWW_AUTHENTICATE, KerberosAuthenticator.NEGOTIATE);
+        Mockito.verify(response, times(2)).setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+    }
+
+    @Test
+    public void testRequestWithInvalidAuthorization() throws Exception {
```
[the remark this answers is not in the document]
Comment (guruchai) [reply common-auth/src/test/java/com/hortonworks/registries/auth/server/TestKerberosBasicAuthenticationHandler.java]: Not exactly. testRequestWithInvalidAuthorization tests the invalidity of Authorization header. testRequestWithInvalidCredentials tests the invalidity of the credentials after validating the Authorization header. They are like 2 steps in parsing step.

**Comment by guruchai** [comment]
Thanks for the review, @kamalcph!! made the changes suggested.

File: common-auth/src/test/java/com/hortonworks/registries/auth/server/TestKerberosBasicAuthenticationHandler.java [hunk 3]
```diff
@@ -0,0 +1,302 @@
+package com.hortonworks.registries.auth.server;
```
Comment (kamalcph) [inline common-auth/src/test/java/com/hortonworks/registries/auth/server/TestKerberosBasicAuthenticationHandler.java]: license is missing.

File: common-auth/src/test/java/com/hortonworks/registries/auth/server/TestKerberosBasicAuthenticationHandler.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kamalcph) [reply common-auth/src/test/java/com/hortonworks/registries/auth/server/TestKerberosBasicAuthenticationHandler.java]: Agree.

File: common-auth/src/main/java/com/hortonworks/registries/auth/util/Utils.java [hunk 4]
```diff
@@ -76,4 +76,21 @@ public static boolean isBrowser(String[] nonBrowserUserAgents, HttpServletReques
...
+    public static boolean isBlank(CharSequence cs) {
+        int strLen;
+        if (cs != null && (strLen = cs.length()) != 0) {
+            for(int i = 0; i < strLen; ++i) {
+                if (!Character.isWhitespace(cs.charAt(i))) {
+                    return false;
+                }
+            }
+            return true;
+        } else {
```
Comment (kamalcph) [inline common-auth/src/main/java/com/hortonworks/registries/auth/util/Utils.java]: `else` block can be removed.

File: common-auth/src/main/java/com/hortonworks/registries/auth/server/KerberosBasicAuthenticationHandler.java [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+ *
+ * <p>
+ * The supported configuration properties that are required in addition to those required by KerberosAuthenticationHandler are:
+ * <ul>
+ * <li>login.enabled: a boolean string to indicate whether the enabling of Kerberos login.</li>
+ * </ul>
+ */
+public class KerberosBasicAuthenticationHandler extends KerberosAuthenticationHandler {
+
+    public static final String LOGIN_ENABLED_CONFIG = "login.enabled";
```
Comment (raju-saravanan) [inline common-auth/src/main/java/com/hortonworks/registries/auth/server/KerberosBasicAuthenticationHandler.java]: Can you mention these props in the registry .yaml under the "servletFilters" section?

File: common-auth/src/main/java/com/hortonworks/registries/auth/server/KerberosBasicAuthenticationHandler.java [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+    public static final String SPNEGO_ENABLED_CONFIG = "spnego.enabled";
+    public static final String TYPE = "kerberos-login";
+    public static final String AUTHORIZATION_HEADER = "Authorization";
+    public static final String BASIC_AUTHENTICATION = "Basic";
+
+    private static final Logger LOG = LoggerFactory.getLogger(KerberosBasicAuthenticationHandler.class);
+    private static final String METHOD = "POST";
```
Comment (raju-saravanan) [inline common-auth/src/main/java/com/hortonworks/registries/auth/server/KerberosBasicAuthenticationHandler.java]: Can we rename this to "HTTP_LOGIN_METHOD" or more something meaningful?

File: common-auth/src/main/java/com/hortonworks/registries/auth/server/KerberosBasicAuthenticationHandler.java [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+    public void init(Properties config) throws ServletException {
+        spnegoEnabled = Boolean.parseBoolean(config.getProperty(SPNEGO_ENABLED_CONFIG, Boolean.TRUE.toString()));
+        if (spnegoEnabled) {
+            super.init(config);
+        }
+
+        try {
+            provider = new KerberosAuthenticationProvider();
+            SunJaasKerberosClient client = new SunJaasKerberosClient();
+            client.setDebug(true);
```
Comment (raju-saravanan) [inline common-auth/src/main/java/com/hortonworks/registries/auth/server/KerberosBasicAuthenticationHandler.java]: I am not sure how much logs the client will generate in debug mode. Can we enable this if debug mode is enabled for LOG?

File: common-auth/src/main/java/com/hortonworks/registries/auth/server/KerberosBasicAuthenticationHandler.java [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Kerberos Login is not attempted because method: {}, secure: {}, authorization is empty: {}", request.getMethod(),
+                        request.isSecure(), authorization.isEmpty());
+            }
+            response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+            return null;
+        }
+        String credentials = authorization.split(BASIC_AUTHENTICATION)[1].trim();
+        byte[] credentialsArray = Base64.getDecoder().decode(credentials);
+        String[] userPassword = new String(credentialsArray, StandardCharsets.UTF_8).split(":");
```
Comment (raju-saravanan) [inline common-auth/src/main/java/com/hortonworks/registries/auth/server/KerberosBasicAuthenticationHandler.java]: Can we rename this to "principalAndPassword"? I was bit confused when rawPrincipal was extracted from "userPassword".

File: common-auth/src/main/java/com/hortonworks/registries/auth/server/KerberosBasicAuthenticationHandler.java [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+            // otherwise, use the given principal, which will use the default realm as specified in the krb5 configuration
+            identity = rawPrincipal;
+            LOG.debug("Realm was not specified in principal {}, default realm is blank and was not added to the identity being authenticated",
+                    rawPrincipal);
+        }
+
+        return identity;
+    }
+
+    static class KerberosUserDetailsService implements UserDetailsService {
```
Comment (raju-saravanan) [inline common-auth/src/main/java/com/hortonworks/registries/auth/server/KerberosBasicAuthenticationHandler.java]: Do we need static keyword here?

File: common-auth/src/main/java/com/hortonworks/registries/auth/server/KerberosBasicAuthenticationHandler.java [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+ * mechanism. If a user provides user credentials in a HTTPS, POST call, then a Kerberos login is attempted. In the authentication failure scenario,
+ * the SPNEGO sequence is invoked.
+ *
+ * <p>
+ * The supported configuration properties that are required in addition to those required by KerberosAuthenticationHandler are:
+ * <ul>
+ * <li>login.enabled: a boolean string to indicate whether the enabling of Kerberos login.</li>
+ * </ul>
+ */
+public class KerberosBasicAuthenticationHandler extends KerberosAuthenticationHandler {
```
Comment (raju-saravanan) [inline common-auth/src/main/java/com/hortonworks/registries/auth/server/KerberosBasicAuthenticationHandler.java]: Can you talk about this authentication method in security.rst under docs pacakge ?

**Comment by guruchai** [comment]
Made the changes.
@raju-saravanan , @kamalcph, can we merge the PR if things look good?

File: common-auth/src/main/java/com/hortonworks/registries/auth/server/KerberosBasicAuthenticationHandler.java (hunk 9 above)
Comment (kamalcph) [reply common-auth/src/main/java/com/hortonworks/registries/auth/server/KerberosBasicAuthenticationHandler.java]: static nested classes are usually preferred if the class is not using any of the fields from the enclosing class.

Comment (guruchai) [reply common-auth/src/main/java/com/hortonworks/registries/auth/server/KerberosBasicAuthenticationHandler.java]: Agreed. I hurried up. Did not pay much attention. Considering no access to enclosing class, static should have been finer for this case!

**Pull request merged**